034-instanced-pipelines/proposal.md

```diff
@@ -0,0 +1,74 @@
+# Summary
+
+Instanced pipelines group together pipelines which share a common template configured with different ((vars)). They provide a simple two-level hierarchy and automatic archiving of instances which are no longer needed.
+
+# Proposal
+
+Pipelines can be configured with 'instance vars' like so:
+
+```yaml
+fly set-pipeline -p branch --instance-var branch=feature/foo
+```
+
+This will configure a `branch` pipeline, with the `((branch))` var set to `"feature/foo"`.
+
+Instance vars are used as part of the pipeline identifier in the UI and API. There can be multiple instances of a pipeline with the same name:
+
+```yaml
+fly set-pipeline -p branch --instance-var branch=feature/foo
+fly set-pipeline -p branch --instance-var branch=feature/bar
+```
+
+Instanced pipelines sharing the same name will be grouped together in the web UI.
+
+An individual instance of a pipeline can be manually destroyed, paused, and archived ([RFC #33](https://github.com/concourse/rfcs/pull/33)):
+
+```yaml
+fly destroy-pipeline -p branch -i branch:feature/foo
+fly pause-pipeline   -p branch -i branch:feature/foo
+fly archive-pipeline -p branch -i branch:feature/foo
+```
+
+(Side note: `:` vs. `=` is a little weird but it's consistent with `fly check-resource` - we use `=` for assignment and `:` for partial filtering.)
+
+## Automatic archival
+
+Instanced pipelines build on the automatic pipeline archiving introduced in [RFC #33](https://github.com/concourse/rfcs/pull/33). Individual pipeline instances that are no longer configured will be automatically archived in the same way that normal pipelines would.
```

Comment:
I'm confused about how to determine "individual pipeline instance no longer configured". My use case is like:

```yaml
- get: policy # this will generate a pipeline yaml on the fly
  trigger: true
- load_var: id
  file: policy/id.txt
- set_pipeline: some-name
  file: policy/pipeline.yaml
  instance_vars:
      id: ((.:id))
```

In this case, `id` is dynamically generated, after build 2, will pipeline set by build 1 be automatically archived?

Reply:
Yep. This is more a behavior of the `set_pipeline` step in general than with instanced pipelines in particular though.

If the previous pipeline should still exist, that must be expressed by the build continuing to run `set_pipeline` against it.

For example, pull request pipeline automation will continuously re-configure all pipelines for all PRs that are open, rather than setting one-pipeline-per-build:

```yaml
set_pipeline: pr
instance_vars: {pr_number: ((.:pr.number))}
matrix: {pr: concourse-prs}
```

Just guessing from your example, it seems like it should be configuring a matrix of pipelines across the valid policy IDs, similar to above.

Reply:
Cool. Thanks for the explanation. Looking forward to this feature.